src/Theme.php

```diff
@@ -0,0 +1,205 @@
+ */
+
+namespace Glpi;
+
+/**
+ * Class that represents a theme/palette.
+ */
+final class Theme
+{
+    private string $key;
+    private string $name;
+    private bool $is_dark;
+    private bool $is_custom;
+
+    public const DEFAULT_THEME = 'auror';
+    public const CORE_THEME_ROOT = GLPI_ROOT . '/css/palettes/';
+
+    /**
+     * Get all themes added by GLPI itself
+     * @return Theme[]
+     */
+    public static function getCoreThemes(): array
+    {
+        static $result = null;
+        if ($result === null) {
+            $result = [
+                new self('aerialgreen', _x('theme', 'Aerial Green'), false, false),
+                new self('auror', _x('theme', 'Auror'), false, false),
+                new self('auror_dark', _x('theme', 'Dark Auror'), true, false),
+                new self('automn', _x('theme', 'Autumn'), false, false),
+                new self('classic', _x('theme', 'Classic'), false, false),
+                new self('clockworkorange', _x('theme', 'Clockwork Orange'), false, false),
+                new self('dark', _x('theme', 'Dark'), false, false),
+                new self('darker', _x('theme', 'Darker'), true, false),
+                new self('flood', _x('theme', 'Flood'), false, false),
+                new self('greenflat', _x('theme', 'Green Flat'), false, false),
+                new self('hipster', _x('theme', 'Hipster'), false, false),
+                new self('icecream', _x('theme', 'Ice Cream'), false, false),
+                new self('lightblue', _x('theme', 'Light Blue'), false, false),
+                new self('midnight', _x('theme', 'Midnight'), true, false),
+                new self('premiumred', _x('theme', 'Premium Red'), false, false),
+                new self('purplehaze', _x('theme', 'Purple Haze'), false, false),
+                new self('teclib', _x('theme', 'Teclib'), false, false),
+                new self('vintage', _x('theme', 'Vintage'), false, false),
+            ];
+        }
+
+        return $result;
+    }
+
+    /**
+     * Get all themes present in the "palettes" directory that aren't core themes
+     * @return Theme[]
+     */
+    public static function getCustomThemes(): array
+    {
+        static $custom_themes = null;
+
+        if ($custom_themes === null) {
+            $custom_themes = [];
+
+            $file_matches = [];
+            // Cannot use GLOB_BRACE on some platforms (like the docker environment used for tests)
+            $patterns = [
+                '*.css',
+                '*.scss'
+            ];
+            foreach ($patterns as $pattern) {
+                foreach (glob(self::CORE_THEME_ROOT . $pattern) as $file) {
+                    $file_name = pathinfo($file, PATHINFO_FILENAME);
+                    if (str_starts_with($file_name, '_')) {
+                        continue;
+                    }
+                    $file_matches[$file_name] = $file;
+                }
+            }
```

Comment:
If we want to officially support custom themes, I think with you use one of the following solutions:
1. Adding a hook that would permit to any plugin to register one or more themes (and so the plugin would be itself able to declare the `$is_dark` value).
2. Using a new `GLPI_THEMES_DIR` constant (with default value to `_files/themes`), and parse this directory.

I personally prefer the first solution.

Reply:
For the first solution, I don't know how approachable it would be for users. IT staff may not necessarily know PHP or be familiar with creating a plugin or want to maintain one (even small changes). I would be nice to have it available as an option just in case someone wants to maintain some kind of theme pack, but I don't think it is good as the only solution.

The second solution more closely matches what it seems like users are already doing which is creating their own CSS files and dumping them in the `css/palettes` folder. At least with them being in `_files/themes` they won't conflict and are clearly user-defined.

With the new UI, creating new themes can be very simple with just copying an existing one and changing color values as needed.

Reply:
First solution is not so complicated. It would not require much than following code.

```php
<?php
function plugin_init_mycompany() {
    global $PLUGIN_HOOKS;
    $PLUGIN_HOOKS['custom_theme']['mycompany'] = [
        // key pattern: `plugin:${plugin_name}/${css_basename}`
        // will search for (s)css file in `css` directory of plugin:
        // - try to load `mycompany/css/mytheme.scss`
        // - fallback to `mycompany/css/mytheme.css` 
        return new \Glpi\UI\Theme('plugin:mycompany/mytheme', 'My company theme', true, true);
    ];
}

function plugin_version_mycompany() {
   return [
      'name'           => 'My company plugin',
      'version'        => '1.0.0',
   ];
}
```

And documenting it would not be more complicated than documenting the other solution.

@orthagh @trasher Which solution would you prefer ?

Reply:
i agree with @cconard96, php code, even minimal, can be a step too much complex for some users.
I think the `GLPI_THEMES_DIR`|`_files/themes` is a requirements.

Both solutions could be implemented but to be sure, what's the advantage to declare it by a plugin hook ? (fixing the is_dark issue only ?)

Reply:
The plugin hook could be useful if their file name doesn't match what is expected for the light/dark check or if they want a different display name for the palette than what gets computed from the file name.
Although, I think both are edge cases.

It could also be used to group similar palettes together and/or distribute them via the existing plugin/marketplace system.

Reply:
> i agree with @cconard96, php code, even minimal, can be a step too much complex for some users.

For users that have no technical background, it is still possible to use the entity configuration to override the CSS.

> I think the `GLPI_THEMES_DIR`|`_files/themes` is a requirement.

Indeed, if we want to legacy behaviour (i.e. custom themes files that have been put in `css` directory), then we should probably trigger a warning to ensure that administrator will move them in the `GLPI_THEMES_DIR`.

> Both solutions could be implemented but to be sure, what's the advantage to declare it by a plugin hook ?

Main advantage is that it would be possible to develop theme packs as plugins that could be available through marketplace. It would then be possible to use the GLPI<>plugin compatibility logic to ensure that theme pack is downloaded in the correct version.

Another minimal advantage is to be able to have a translated name, and also, as you said, to have no ambiguity about the light/dark mode.